components/core/src/package/target.rs

```diff
@@ -0,0 +1,197 @@
+// limitations under the License.
+
+use std::fmt;
+use std::result;
+use std::str::FromStr;
+use os::system::{uname, Uname};
+
+use error::{Error, Result};
+
+pub trait Target: fmt::Display + Into<PackageTarget> {
+    fn platform(&self) -> &str;
+    fn architecture(&self) -> &str;
+    fn valid_for(&self, uname: Uname) -> Result<()>;
+}
+
+/// Describes the platform (operating system/kernel)
+/// and architecture (x86_64, i386, etc..) that a package is built for
+#[derive(RustcEncodable, RustcDecodable, Debug, Clone, Hash)]
+pub struct PackageTarget {
+    pub platform: String,
+    pub architecture: String,
+}
+
+impl PackageTarget {
+    /// Creates a new package target
+    ///
+    /// Errors:
+    /// * InvalidPackageTargetPlatform
+    /// * InvalidPackageTargetArchitecture
+    pub fn new<T: Into<String>>(platform: T, architecture: T) -> Self {
+        let valid_platform = PackageTarget::supported_platform(platform).unwrap();
+        let valid_arch = PackageTarget::supported_architecture(architecture).unwrap();
+        PackageTarget {
+            platform: valid_platform,
+            architecture: valid_arch,
+        }
+    }
+
+    fn supported_platform<T: Into<String>>(value: T) -> Result<String> {
+        let platform = value.into();
+        let valid_platforms = vec!["linux", "windows",];
+        if valid_platforms.iter().any(|&x| x == platform) {
+            Ok(platform)
+        } else {
+            Err(Error::InvalidPackageTargetPlatform(platform))
+        }
+    }
+
+    fn supported_architecture<T: Into<String>>(value: T) -> Result<String> {
+        let architecture = value.into();
+        let valid_architectures = vec!["x86_64"];
+        if valid_architectures.iter().any(|&x| x == architecture) {
+            Ok(architecture)
+        } else {
+            Err(Error::InvalidPackageTargetArchitecture(architecture))
+        }
+    }
+}
+
+impl Target for PackageTarget {
+    fn platform(&self) -> &str {
+        &self.platform
+    }
+    fn architecture(&self) -> &str {
+        &self.architecture
+    }
+    fn valid_for(&self, uname: Uname) -> Result<()> {
+        let architecture = uname.machine.trim().to_lowercase();
+        let platform = uname.sys_name.trim().to_lowercase();
+        if self.architecture == architecture && self.platform == platform {
+            Ok(())
+        } else {
+            Err(Error::TargetMatchError(format!("Package target ({}-{}) does not match system \
+                                                 target ({}-{}).",
+                                                self.architecture,
+                                                self.platform,
+                                                architecture,
+                                                platform,)))
+        }
+    }
+}
+
+impl Default for PackageTarget {
+    fn default() -> PackageTarget {
+        let current_system = uname().unwrap();
```

Comment:
This will cause a panic without message in a default implementation which is a bit of a no-no in the API. Is there a "safer" way to represent a default if we can't get the uname or perhaps we do want to panic but we should provide a bit of a message here.

`expect()` is just like `unwrap()` but it allows you to specify a panic message